File: cases/taint/scala/alias-propagation-positive/AliasPropagationPositive.scala
package dataflowbench

object AliasPropagationPositive {
  class Holder {
    var value: Int = 0
  }

  def dfb_source(): Int = { // DFB-SOURCE: alias-propagation-input
    1
  }

  def dfb_sink(value: Int): Unit = {} // DFB-SINK: alias-propagation-sink

  def run(): Unit = {
    val original = new Holder()
    val alias = original // DFB-WITNESS: alias-propagation-alias
    val distinct = new Holder()
    original.value = dfb_source() // DFB-WITNESS: alias-propagation-store
    dfb_sink(alias.value)
  }
}

File: cases/taint/c/alias-propagation-positive/alias_propagation_positive.c
struct Holder {
    int value;
};

int dfb_source(void) { // DFB-SOURCE: alias-propagation-input
    return 1;
}

void dfb_sink(int value) {} // DFB-SINK: alias-propagation-sink

void run(void) {
    struct Holder original;
    struct Holder *alias = &original; // DFB-WITNESS: alias-propagation-alias
    struct Holder distinct;
    original.value = dfb_source(); // DFB-WITNESS: alias-propagation-store
    distinct.value = 0;
    dfb_sink(alias->value);
}
